### PR TITLE
Cts prevent removing sinks

### PR DESCRIPTION
**Pull request by arthurjolo**
During the HTree building step of cts we can create more Clock subNets then needed, these extra Clock subNets were deleted at the end of CTS, but this can cause crashes (like it was mentioned in [PR#4938](https://github.com/The-OpenROAD-Project/OpenROAD/pull/4938)).

To prevent having to delete these nets at the end of CTS, this PR identifies and removes theses nets during the HTree building step.

### CODE REVIEW

File: src/cts/src/Clock.h [hunk 1]
```diff
@@ -39,6 +39,7 @@
 #include <cstdint>
 #include <deque>
 #include <functional>
+#include <iostream>
```
Comment (eder-matheus) [inline src/cts/src/Clock.h:42]: Do you need this include? I think you're not using nothing from iostream

File: src/cts/src/Clock.h [hunk 2]
```diff
@@ -218,6 +223,44 @@ class Clock
     return subNets_.back();
   }
 
+  void removeSubNet(ClockSubNet sub_net)
+  {
+    std::set<ClockInst*> sinksToRemove;
+    sinksToRemove.insert(sub_net.getDriver());
+    for (auto it = subNets_.begin(); it != subNets_.end();) {
```
Comment (eder-matheus) [inline src/cts/src/Clock.h:230]: What is the cost of iterating over every subnet? Wouldn't the runtime of it be too large?

File: src/cts/src/Clock.h [hunk 3]
```diff
@@ -218,6 +223,44 @@ class Clock
...
+      } else {
+        ++it;
+      }
+    }
+    removeClockBuffer(sub_net.getDriver());
+  }
+
+  void removeClockBuffer(ClockInst* clock_buffer)
+  {
+    for (auto it = clockBuffers_.begin(); it != clockBuffers_.end();) {
```
Comment (eder-matheus) [inline src/cts/src/Clock.h:254]: Similar question here. Isn't there a way to avoid these loops? Perhaps a hash table, a map...

**Review by eder-matheus — CHANGES_REQUESTED** [review]

File: src/cts/src/Clock.h (hunk 2 above)
Comment (arthurjolo) [reply src/cts/src/Clock.h:230]: The run time is O(n) for the number of subNets, I don't think there is a better runtime for removing this an element from a std::deque. Also this code shouldn't be run very often, it is to fix some corner cases. But if the runtime becomes a problem we can maintain the subNet deletion at the end of CTS that would not affect the runtime.

**Comment by arthurjolo** [comment]
This PR generated a crash in one testcases, and the approach to solve the problem had to change. So I will be closing this PR and open a new one with the correct fix.

**Pull request closed without merging**